File: js/trie.js
// Trie 

class Trie {
    
    constructor(mx, letter) {
        this.root = 1; 
        this.unused = 2; 
        this.mx = mx // the maximum number of input. 
        this.chk = Array(mx).fill(false); 
        this.nxt = Array.from({ length: mx}, () => Array(letter).fill(-1)); 
    }

    convert(letter) {
        return letter.charCodeAt(0) - 'a'.charCodeAt(0);
    }

    insert(word) {
        // word is string. 
        let curIdx = this.root; 

        for (let i = 0; i < word.length; i++) {
            let curLetter = this.convert(word[i]); 
            let nxt = this.nxt[curIdx][curLetter]; 
            if (nxt === -1) {
                this.nxt[curIdx][curLetter] = this.unused++; 
            }
            curIdx = this.nxt[curIdx][curLetter]; 
        }

        this.chk[curIdx] = true; 
    }

    erase(word) {
        // success returns true, failure returns false. 
        let curIdx = this.root; 

        for (let i = 0; i < word.length; i++) {
            let curLetter = this.convert(word[i]); 
            let nxt = this.nxt[curIdx][curLetter]; 
            if (nxt === -1) {
                return false; 
            }
            curIdx = this.nxt[curIdx][curLetter]; 
        }

        if (this.chk[curIdx] === true) {
            this.chk[curIdx] = false; 
            return true; 
        } else {
            return false; 
        }
    }

    find(word) {
        // success returns true, failure returns false. 
        let curIdx = this.root; 

        for (let i = 0; i < word.length; i++) {
            let curLetter = this.convert(word[i]); 
            let nxt = this.nxt[curIdx][curLetter]; 
            if (nxt === -1) {
                return false; 
            }
            curIdx = this.nxt[curIdx][curLetter]; 
        }
        return this.chk[curIdx] === true; 
    }
}

File: cpp/trie.cpp
#include <bits/stdc++.h>
using namespace std;

class Trie {

public:
    const int ROOT = 1;
    int unused = 2;
    int MX;
    vector<bool> chk;
    vector<vector<int>> nxt; 

    Trie(int _MX): MX(_MX) {
        chk.assign(MX, false);

        nxt.assign(MX, vector<int>{});
        for (int i = 0; i < MX; i++) {
            nxt[i].push_back(-1); 
        }
    }

    ~Trie() = default;
    
    int c2i(char c) {
        return c - 'a'; 
    }

    void insert(string &s) {
        int cur = ROOT; 
        for (auto c : s) {
            if(nxt[cur][c2i(c)] == -1) {
                nxt[cur][c2i(c)] = unused++; 
            }
            cur = nxt[cur][c2i(c)]; 
        }
        chk[cur] = true; 
    }

    void erase(string &s) {
        int cur = ROOT; 
        for (auto c: s) {
            if (nxt[cur][c2i(c)] == -1) {
                return; 
            }
            cur = nxt[cur][c2i(c)]; 
        }
        chk[cur] = false; 
    }

    bool find(string &s) {
        int cur = ROOT; 
        for (auto c : s) {
            if (nxt[cur][c2i(c)] == -1) {
                return false; 
            }
            cur = nxt[cur][c2i(c)]; 
        }
        return chk[cur]; 
    }
};